flecsi/exec/hpx/task_prologue.hh

```diff
@@ -0,0 +1,223 @@
+    // it has to be used a dependency for executing the current task
+    auto & field_future = r.get_future(f);
+    if(field_future.valid()) {
+      std::lock_guard l(spinlock_pool::spinlock_for(&field_future));
+      dependencies.push_back(field_future);
+    }
+
+    const auto storage = r.template get_storage<T, ProcessorType>(f);
+    if constexpr(glob) {
+      using data_type = ::hpx::serialization::serialize_buffer<T>;
+      auto comm = flecsi::run::context::instance().world_comm(
+        "prolog_visit" + std::to_string(f));
+      if(reg.ghost<privilege_pack<get_privilege(0, P), ro>>(f)) {
+        if(comm.is_root()) {
+          auto f = ::hpx::collectives::broadcast_to(comm,
+            data_type(storage.data(), storage.size(), data_type::reference));
+          f.get();
+        }
+        else {
+          auto f = ::hpx::collectives::broadcast_from<data_type>(comm);
+          auto && data = f.get();
+          assert(data.size() == storage.size());
+          std::move(data.begin(), data.begin() + data.size(), storage.data());
+        }
+      }
+    }
+    else {
+      reg.ghost_copy<P>(ref);
+    }
+
+    if(privilege_write(P)) {
+      // request future from promise only if required
+      if(!future.valid()) {
+        future = promise.get_future();
+      }
+      // if the task writes to the current argument then we must associate
+      // the future that represents the result of the task with the argument
+      std::lock_guard l(spinlock_pool::spinlock_for(&field_future));
+      field_future = future;
+      has_dependencies = true;
+    }
+
+    accessor.bind(storage);
+  } // visit generic topology
+
+public:
+  /*!
+    Delay the execution of the given task until all dependencies have been
+    satisfied.
+   */
+  template<typename R, typename Params, typename Task>
+  auto delay_execution(Params && params, std::string task_name, Task && task) {
+
+    if(dependencies.empty()) {
+      // TIP: param_buffers is an RAII type. We create an instance and give
+      // the object a reference to the parameters and name of the task. We
+      // then assign it to the `finalize` variable so it is not destroyed
+      // immediately. The object will be destroyed when reduce_internal()
+      // returns. The ~param_buffers() will then perform the necessary clean
+      // up on the parameters (mostly calling mutator.commit()).
+      auto finalize = param_buffers(params, task_name);
+
+      // directly execute task as no dependencies have to be taken into account
+      if constexpr(std::is_void_v<R>) {
+        task(std::forward<Params>(params), std::move(task_name));
+        return hpx::make_ready_future();
+      }
+      else {
+        return hpx::make_ready_future(
+          task(std::forward<Params>(params), std::move(task_name)));
+      }
+    }
+
+    // delay execution to start only after all dependencies have been satisfied
+    return ::hpx::dataflow(
+      ::hpx::launch::sync,
+      [&,
```

Comment:
It seems that `task` (and not `this`) is the only thing being captured here.  It's a temporary created in `reduce_internal`; does it live long enough to use it here?